### PR TITLE
Xml2xml

### PR DESCRIPTION
**Pull request by epifanio**
This PR does a refactoring of the code to work with XML files. The code is as general as possible so that some of the functionalities can be applied not just to MMD and/or ISO but to any XML (hence the renaming of methods and files)
This P.R. also includes a new script to ruin transformation in parallel and adds few more tests, so to reach full coverage.

```nosetests --with-coverage --cover-package=py_mmd_tools```

```
................
Name                       Stmts   Miss  Cover   Missing
--------------------------------------------------------
py_mmd_tools/__init__.py       0      0   100%
py_mmd_tools/log_util.py      31      0   100%
py_mmd_tools/xml2xml.py       28      0   100%
py_mmd_tools/xml_util.py      24      0   100%
--------------------------------------------------------
TOTAL                         83      0   100%
----------------------------------------------------------------------
Ran 16 tests in 30.550s

OK
```
other minor changes in this P.R.:
* It is using the new data directory to load the test dataset and removes the dependency from the `confuse` 
* the logging utility methods accept logging to file or as std output - to avoid issue when running in parallel.

### CODE REVIEW

**Review by ElodieFZ — APPROVED** [review]

File: py_mmd_tools/xml2xml.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import pathlib
+import time
+
+import lxml.etree as ET
+from py_mmd_tools.xml_util import xml_check, xsd_check
+import errno
+import os
+
+def xml2xml(
```
Comment (mortenwh) [inline py_mmd_tools/xml2xml.py:16]: `xml_translate`

File: py_mmd_tools/xml2xml.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    xml_file,
+    outputfile,
+    xslt,
+    xsd_validation=False,
+    xsd_schema=None,
+):
+    """[Transform MMD file to ISO using xslt]
```
Comment (mortenwh) [inline py_mmd_tools/xml2xml.py:23]: needs to be updated

File: py_mmd_tools/log_util.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+"""
+Tool for converting metadata from MMD format to ISO format using a specific xslt.
+ License:
+     This file is part of the S-ENDA-Prototype repository (https://github.com/metno/py-mmd-tools).
```
Comment (mortenwh) [inline py_mmd_tools/log_util.py:4]: Should be: `License: This file is part of py-mmd-tools, licensed under the Apache License 2.0 (https://www.apache.org/licenses/LICENSE-2.0)`

File: py_mmd_tools/log_util.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+     S-ENDA-Prototype is licensed under GPL-3.0 (https://github.com/metno/py-mmd-tools/blob/master/LICENSE)
+"""
+
+import logging
+import pathlib
+import os
+
+def get_logpath(logdirpath):
+    if not pathlib.Path(logdirpath).exists():
```
Comment (mortenwh) [inline py_mmd_tools/log_util.py:13]: Please add a docstring

File: py_mmd_tools/log_util.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        except IOError:
+            raise
+    else:
+        if os.access(logdirpath, os.W_OK):
+            return logdirpath
+        else:
+            raise IOError
+
+
+def setup_log(name, logdirpath, logtype='stream'):
```
Comment (mortenwh) [inline py_mmd_tools/log_util.py:26]: please add docstring

File: py_mmd_tools/log_util.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
+"""
+Tool for converting metadata from MMD format to ISO format using a specific xslt.
```
Comment (mortenwh) [inline py_mmd_tools/log_util.py:2]: Seems this file is for something else?

File: py_mmd_tools/log_util.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        fh.setFormatter(log_format)
+        logger.addHandler(fh)
+        return logger
+    if logtype=='stream':
+        # streamhandler
+        ch = logging.StreamHandler()
+        ch.setLevel(logging.INFO)
+        ch.setFormatter(log_format)
+        logger.addHandler(ch)
+        return logger
```
Comment (mortenwh) [inline py_mmd_tools/log_util.py:46]: Repetition - can you unindent `return logger`, and delete line 39?

File: py_mmd_tools/xml2xml.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
+"""
+Tool for converting metadata from MMD format to ISO format using a specific xslt.
```
Comment (mortenwh) [inline py_mmd_tools/xml2xml.py:2]: This is now a generic tool, isn't it? Can you update the docstring?

File: py_mmd_tools/xml2xml.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), xml_file)
+        
+    if xsd_validation:
+        if xsd_schema is None:
+            raise TypeError
+        if not pathlib.Path(xsd_schema).exists():
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), xsd_schema)
+        else: 
+            if not xsd_check(xml_file, xsd_schema=xsd_schema):
+                return False
```
Comment (mortenwh) [inline py_mmd_tools/xml2xml.py:44]: you have three lines that return either True or False. Wouldn't it be better to assign the result to a variable, and return it in the end of the method?

File: py_mmd_tools/xml_util.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
+"""
+Utility tool to work on mmd xml files.
+ License:
+     This file is part of the S-ENDA-Prototype repository (https://github.com/metno/py-mmd-tools).
```
Comment (mortenwh) [inline py_mmd_tools/xml_util.py:4]: Please change to License: `This file is part of py-mmd-tools, licensed under the Apache License 2.0 (https://www.apache.org/licenses/LICENSE-2.0)`

File: py_mmd_tools/xml_util.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    """[validate xml syntax from filepath]
+    Args:
+        xml_file ([str]): [filepath to an xml file]
+    Returns:
+        [bool]: [return True if a valid xml filepath is provided, 
+        raises an exception if the xmlfile is invalid, empty, or doesn't exist ]
+    """
+    if pathlib.Path(xml_file).is_file() and os.path.getsize(xml_file) != 0:
+        try:
+            xml = ET.parse(xml_file)
```
Comment (mortenwh) [inline py_mmd_tools/xml_util.py:27]: isn't it enough with this line? FileNotFound and XMLSyntaxError should be raised by `ET.parse`, shouldn't it?

File: py_mmd_tools/xml_util.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+def xsd_check(xml_file, xsd_schema=None):
+    """[validate xml file from filepath]
+    Args:
+        xmlfile ([str]): [filepath to an mmd xml file]
+        xsd_schema ([str]): [filepath to an xsd schema file]
+    Returns:
+        [bool]: [return True if a valid xml filepath is provided, 
+        return False if the xmlfile is invalid, empty, or doesn't exist ]
+    """
+    if not pathlib.Path(xsd_schema).is_file() and os.path.getsize(xsd_schema) != 0:
```
Comment (mortenwh) [inline py_mmd_tools/xml_util.py:49]: Seems this and the next three lines are obsolete? Don't lines 53-55 raise the correct exceptions?

File: py_mmd_tools/xml_util.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), xml_file)
+    else:
+        if pathlib.Path(xsd_schema).is_file() and xml_check(xsd_schema):
+            xmlschema_mmd = ET.XMLSchema(ET.parse(xsd_schema))
+            xml_doc = ET.ElementTree(file=xml_file)
+            if not xmlschema_mmd.validate(xml_doc):
+                return False
```
Comment (mortenwh) [inline py_mmd_tools/xml_util.py:56]: I prefer one return statement, and a variable assignment..

File: script/xmlconverter.py [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+import sys
+import parmap
+from lxml.etree import XMLSyntaxError
+from xml2xml import xml2xml
+
+"""
+Script to run the xmlc onversion tools
+
+ License:
+     This file is part of the S-ENDA-Prototype repository (https://github.com/metno/py-mmd-tools).
```
Comment (mortenwh) [inline script/xmlconverter.py:13]: please change to `License: This file is part of py-mmd-tools, licensed under the Apache License 2.0 (https://www.apache.org/licenses/LICENSE-2.0)`

File: script/xmlconverter.py [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    xml_files = []
+    for subdir, dirs, files in os.walk(directory):
+        for file in files:
+            # print('File: %s' %file)
+            file_path = subdir + os.sep + file
+            if file_path.endswith(".xml"):
+                xml_files.append(file_path)
+    return xml_files
+
+def writerecord(xml_file, xslt, outdir="/tmp"):
```
Comment (mortenwh) [inline script/xmlconverter.py:43]: perhaps `translate_and_write`? (it does more than just writing the record..)

File: script/xmlconverter.py [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+            xmlfiles,
+            xslt=xslt,
+            outdir=outdir,
+            pm_pbar=False,
+        )
+    else:
+        for i in xmlfiles:
+            try:
+                writerecord(xml_file=i, xslt=xslt, outdir=outdir)
+            except XMLSyntaxError:
```
Comment (mortenwh) [inline script/xmlconverter.py:76]: adding `as e` in this line, and `e.message` in the next might be useful

File: script/xmlconverter.py [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+                print(f"failed on: {i}")
+
+
+def check_record(record, tobedone):
```
Comment (mortenwh) [inline script/xmlconverter.py:80]: please add docstring

File: script/xmlconverter.py [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+    if pathlib.Path(record).stem in tobedone:
+        return record
+
+
+def recover_task(metadata, outdir, parallel=False):
```
Comment (mortenwh) [inline script/xmlconverter.py:85]: docstring needed

File: tests/test_xml_util.py [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+        # unset the output limit when printing the xml diff
+        #
+        current_dir = pathlib.Path.cwd()
+        self.reference_xml = str(current_dir / 'tests' / 'data' / 'reference_mmd.xml')
+        self.reference_xsd = str(current_dir / 'tests' / 'data' / 'mmd.xsd')
+        self.not_a_file = str(current_dir / 'tests' / 'data' / 'not_a_file.xml')
+        self.not_a_valid_xml = str(current_dir / 'tests' / 'data' / 'not_a_valid_xml.xml')
+ 
+    def test_xml_check_1(self):
+        unittest.TestCase.assertEqual(self, xml_check(xml_file=self.reference_xml),
```
Comment (mortenwh) [inline tests/test_xml_util.py:24]: `self.assertEqual(xml_check(xml_file=self.reference_xml), True)` -  isn't this working?

Or even better: `self.assertTrue(xml_check(xml_file=self.reference_xml))`

**Review by mortenwh — COMMENTED** [review]
Please check inline comments. There's a few changes that needs to be done.

**Comment by epifanio** [comment]
I addressed most of your comments. Making a new PR with the changes - Note that the proposed changes affected code coverage a bit - I didn't manage to get it higher has it was before.

File: py_mmd_tools/xml_util.py (hunk 11 above)
Comment (epifanio) [reply py_mmd_tools/xml_util.py:27]: I may be wrong, but it looks like `ET.parse` raises a `OSerror`  when a non existent file is given directly, example below.:

```python
In [1]: import lxml.etree as ET  
In [2]: xml = ET.parse('tests/data/this_fine_not_exist.xml')                                                                                                                  
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-7-60df23105c82> in <module>
----> 1 xml = ET.parse('tests/data/this_fine_not_exist.xml')

src/lxml/etree.pyx in lxml.etree.parse()

src/lxml/parser.pxi in lxml.etree._parseDocument()

src/lxml/parser.pxi in lxml.etree._parseDocumentFromURL()

src/lxml/parser.pxi in lxml.etree._parseDocFromFile()

src/lxml/parser.pxi in lxml.etree._BaseParser._parseDocFromFile()

src/lxml/parser.pxi in lxml.etree._ParserContext._handleParseResultDoc()

src/lxml/parser.pxi in lxml.etree._handleParseResult()

src/lxml/parser.pxi in lxml.etree._raiseParseError()

OSError: Error reading file 'tests/data/not_a_valid_xmlh.xml': failed to load external entity "tests/data/this_fine_not_exist.xml"
```

**Pull request merged**